### PR TITLE
initial work on new api

### PR DESCRIPTION
**Pull request by lukewestby**
as per #66

### CODE REVIEW

File: modules/cmd.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        .then((action) => [action])
+
+    case cmdTypes.CONSTANT:
+      return Promise.resolve(cmd.action)
+
+    case cmdTypes.BATCH:
+      const filteredCmds = cmd.cmds
+        .filter((nextCmd) => nextCmd.type !== cmdTypes.NONE)
+
+      const batchedPromises = filteredCmds.map(cmdToPromise).filter((x) => x)
```
Comment (bdwain) [inline modules/cmd.js]: i don't think you need to filter above and then filter here also. seems like the only promise that can be falsey is none. you can probably just say `batchedPromises = cmd.cmds.map(cmdToPromise),filter(x => x)`

Comment (lukewestby) [inline modules/cmd.js]: yeah i'll get rid of the first filter. nested empty batches will also return `null` so both cases need to be handled.

Comment (bdwain) [inline modules/cmd.js]: i think they'll be handled by the line below `if (batchedPromises.length === 0) return null` right?

File: modules/loop.js [hunk 2]
```diff
@@ -1,76 +1,34 @@
...
-      isEffect(effect),
-      'Given effect is not an effect instance.'
-    );
+      isCmd(cmd),
+      'Given cmd is not an Cmd instance.'
+    )
   }
 
-  return [model, effect];
+  return [model, effect]
```
Comment (bdwain) [inline modules/loop.js]: this should be [model, cmd] i beleive

File: modules/install.js [hunk 3]
```diff
@@ -1,67 +1,68 @@
...
+      cmdsQueue.push({ originalAction: action, cmd })
+      return model
+    }
+
+    const store = next(liftReducer(reducer), initialModel, enhancer)
+
+    const runCmds = (queue) => {
+      const promises = queue.map(runCmd).filter((x) => x)
+      if (promises.length === 0) return noCmdPromise
+      else if (promises.length === 1) return promises[0].then(() => {})
```
Comment (bdwain) [inline modules/install.js]: why the empty promise handlers?

File: modules/combineReducers.js [hunk 4]
```diff
@@ -1,55 +1,45 @@
...
-            optimizeBatch(effects)
-        );
-    };
+export const combineReducers = (
+  reducerMap,
+  rootState = {},
+  accessor = defaultAccessor,
+  mutator = defaultMutator
+) => {
+  return const finalReducer = (state = rootState, action) => {
```
Comment (bdwain) [inline modules/combineReducers.js]: error on this line

**Comment by bdwain** [comment]
this looks like all of the cmds have optional action creators. did you mean to do that or do you still want a separate thing where they're optional and have cmds require them?

File: modules/install.js (hunk 3 above)
Comment (lukewestby) [inline modules/install.js]: i'm not sure what the promise should actually resolve at this point so i've left it blank for now

Comment (bdwain) [inline modules/install.js]: got it

**Comment by lukewestby** [comment]
right now it's in a state where we could go either way. i still like the idea of being explicit with arguments but i think something like `Cmd.promise(createPromise, undefined, undefined, ...args)` could be okay as well

**Pull request merged**